librec/src/main/java/librec/util/Sims.java

```diff
@@ -1,228 +1,228 @@
+		}
+
+		double sim = u.size() / sum;
+		if (Double.isInfinite(sim))
+			sim = 1.0;
+
+		return sim;
+	}
+
+	/**
+	 * calculate Pearson Correlation Coefficient (PCC) between two vectors of ratings
+	 * 
+	 * @param a
+	 *            first vector of ratings
+	 * @param b
+	 *            second vector of ratings
+	 * @return Pearson Correlation Coefficient (PCC) value. <br/>
+	 *         If vector a or b is null or the length is less than 2, Double.NaN is returned.
+	 */
+	public static double pcc(List<? extends Number> a, List<? extends Number> b) {
+		if (a == null || b == null || a.size() < 2 || b.size() < 2 || a.size() != b.size())
+			return Double.NaN;
+
+		double mu_a = Stats.mean(a);
+		double mu_b = Stats.mean(b);
+
+		double num = 0.0, den_a = 0.0, den_b = 0.0;
+		for (int i = 0; i < a.size(); i++) {
+			double ai = a.get(i).doubleValue() - mu_a;
+			double bi = b.get(i).doubleValue() - mu_b;
+
+			num += ai * bi;
+			den_a += ai * ai;
+			den_b += bi * bi;
+		}
+
+		return num / (Math.sqrt(den_a) * Math.sqrt(den_b));
+	}
+
+	/**
+	 * calculate extend Jaccard Coefficient between two vectors of ratings
+	 */
+	public static double exJaccard(List<Double> a, List<Double> b) {
+		double num = 0.0, den_a = 0.0, den_b = 0.0;
+		for (int i = 0; i < a.size(); i++) {
+			double ai = a.get(i);
+			double bi = b.get(i);
+
+			num += ai * bi;
+			den_a += ai * ai;
+			den_b += bi * bi;
+		}
+
+		return num / (den_a + den_b - num);
+	}
+
+	/**
+	 * calculate Dice Coefficient between two vectors of ratings
+	 */
+	public static double dice(List<Double> a, List<Double> b) {
+		double num = 0.0, den_a = 0.0, den_b = 0.0;
+		for (int i = 0; i < a.size(); i++) {
+			double ai = a.get(i);
+			double bi = b.get(i);
+
+			num += 2 * ai * bi;
+			den_a += ai * ai;
+			den_b += bi * bi;
+		}
+
+		return num / (den_a + den_b);
+	}
+
+	/**
+	 * Jaccard's coefficient is defined as the number of common rated items of two users divided by the total number of
+	 * their unique rated items.
+	 * 
+	 * @return Jaccard's coefficient
+	 */
+	public static double jaccard(List<Integer> uItems, List<Integer> vItems) {
+		int common = 0, all = 0;
+
+		Set<Integer> items = new HashSet<>();
+		items.addAll(uItems);
+		items.addAll(vItems);
+
+		all = items.size();
+		common = uItems.size() + vItems.size() - all;
+
+		return (common + 0.0) / all;
+
+	}
+
+	/**
+	 * Kendall Rank Correlation Coefficient
+	 * 
+	 * @author Bin Wu
+	 * 
+	 */
+	public static double krcc(List<Double> uItems, List<Double> vItems) {
+		int common = 0, all = 0;
+		double sum = 0;
+		List<Integer> temp = new ArrayList<>();
+
+		Set<Double> items = new HashSet<>();
+		items.addAll(uItems);
+		items.addAll(vItems);
+
+		all = items.size();
+		common = uItems.size() + vItems.size() - all;
+		for (int i = 0; i < uItems.size(); i++) {
+			if (uItems.get(i) > 0 && vItems.get(i) > 0) {
+				temp.add(i);
+			}
+		}
+		for (int m = 0; m < temp.size(); m++) {
+			for (int n = m + 1; n < temp.size(); n++) {

```

Comment:
When counting the number of discordant pairs, the two items should not be the same one. Although it does not matter whether to change this line of code or not, because the product is 0 (not less than 0) if they are the same, I still advise to change it.